src/3d/qgscameracontroller.cpp

```diff
@@ -617,6 +617,25 @@ void QgsCameraController::onPositionChangedTerrainNavigation( Qt3DInput::QMouseE
 
 void QgsCameraController::onPositionChangedGlobeTerrainNavigation( Qt3DInput::QMouseEvent *mouse )
 {
+  const bool hasShift = ( mouse->modifiers() & Qt::ShiftModifier );
+  const bool hasCtrl = ( mouse->modifiers() & Qt::ControlModifier );
+  const bool hasLeftButton = ( mouse->buttons() & Qt::LeftButton );
+  const bool hasMiddleButton = ( mouse->buttons() & Qt::MiddleButton );
+
+  if ( ( hasLeftButton && hasShift && !hasCtrl ) || ( hasMiddleButton && !hasShift && !hasCtrl ) )
+  {
+    setMouseParameters( MouseOperation::RotationCenter, mMousePos );
+
+    float scale = static_cast<float>( std::max( mScene->engine()->size().width(), mScene->engine()->size().height() ) );
+    float pitchDiff = 180.0f * static_cast<float>( mouse->y() - mClickPoint.y() ) / scale;
+    float yawDiff = -180.0f * static_cast<float>( mouse->x() - mClickPoint.x() ) / scale;
+
+    mCameraPose.setPitchAngle( mRotationPitch + pitchDiff );
+    mCameraPose.setHeadingAngle( mRotationYaw + yawDiff );
```

Comment:
Why not keep the double and make only static_cast at the end?
```suggestion
    double scale = std::max( mScene->engine()->size().width(), mScene->engine()->size().height() );
    double pitchDiff = 180.0 * ( mouse->y() - mClickPoint.y() ) / scale;
    double yawDiff = -180.0 * ( mouse->x() - mClickPoint.x() ) / scale;

    mCameraPose.setPitchAngle( static_cast<float>( mRotationPitch + pitchDiff ) );
    mCameraPose.setHeadingAngle( static_cast<float>( mRotationYaw + yawDiff ) );
```

Reply:
Because the pitch/yaw angles are represented as floats in QgsCameraPose, QQuaternion and elsewhere, so there's no point in using doubles and doing extra casts. Floats have more than enough precision for camera angles...